lib/stages/visSharedMemWriter.cpp

```diff
@@ -0,0 +1,340 @@
+        write_to_memory(frame, vis_time_ind_map.at(t), freq_ind);
+        return true;
+    }
+
+    time_ctype max_time = vis_time_ind_map.rbegin()->first;
+    time_ctype min_time = vis_time_ind_map.begin()->first;
+
+    if (t < min_time) {
+        // this data is older than anything else in the map, so we should
+        // just drop it
+        INFO("Dropping integration as buffer (FPGA count: {:d}) arrived too late (minimum in pool {:d})", t.fpga_count, min_time.fpga_count);
+        return false;
+    }
+
+    else if (t > max_time) {
+        // we need to drop the oldest time
+        // and replace it with the new time index
+        cur_pos++;
+        reset_memory(cur_pos);
+        vis_time_ind_map.erase(min_time);
+        vis_time_ind_map[t] = cur_pos;
+        write_to_memory(frame, vis_time_ind_map.at(t), freq_ind);
+        return true;
+    }
+    else {
+        INFO("Dropping integration as buffer (FPGA count: {:d}) arrived too late (only accepting new times greater than {:d})", t.fpga_count, max_time.fpga_count);
+        return false;
+    }
+}
+
+void visSharedMemWriter::reset_memory(uint32_t time_ind) {
+
+    uint8_t *buf_write_pos = buf_addr + (time_ind * nfreq * frame_size);
+    uint64_t *record_write_pos = record_addr + time_ind;
+
+    // each time, we block off an entire time_ind
+
+    if (sem_wait(sem) == -1) {
+        FATAL_ERROR("Failed to acquire semaphore {}", sem_name);
+        return;
+    }
+    memcpy(record_write_pos, &in_progress, sizeof(in_progress));
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to post semaphore {}", sem_name);
+        return;
+    }
+
+    std::vector<char> zeros(nfreq * frame_size, 0);
+    char* tmp = zeros.data();
+    memcpy(buf_write_pos, &tmp, nfreq * frame_size);
+
+    int64_t invalid = -1;
+
+    if (sem_wait(sem) == -1) {
+        FATAL_ERROR("Failed to acquire semaphore {}", sem_name);
+        return;
+    }
+    memcpy(record_write_pos, &invalid, sizeof(invalid));
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to post semaphore {}", sem_name);
+        return;
+    }
+}
+
+void visSharedMemWriter::write_to_memory(const visFrameView& frame, uint32_t time_ind, uint32_t freq_ind) {
+
+    uint8_t *buf_write_pos = buf_addr + ((time_ind * nfreq + freq_ind) * frame_size);
+    uint64_t *record_write_pos = record_addr + time_ind;
+
+    // each time, we block off an entire time_ind
+
+    if (sem_wait(sem) == -1) {
+        FATAL_ERROR("Failed to acquire semaphore {}", sem_name);
+        return;
+    }
+    memcpy(record_write_pos, &in_progress, sizeof(in_progress));
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to release semaphore {}", sem_name);
+        return;
+    }
+
+    memcpy(buf_write_pos + 4, frame.metadata(), metadata_size);
+    memcpy(buf_write_pos + metadata_size + 4, frame.data(), data_size);
+    memcpy(buf_write_pos, &one, 4);
+
+    uint64_t fpga_seq = frame.metadata()->fpga_seq_start;
```

Comment:
Richard: "oops. I said to use this but I was wrong. Rick has pointed out the error of my ways, and I think we should just go back to having an incrementing counter here."

Reply:
Hah!